src/main/java/org/mariadb/jdbc/internal/failover/impl/MastersSlavesListener.java

```diff
@@ -434,80 +434,88 @@ public void lockAndSwitchSecondary(Protocol newSecondaryProtocol) throws QueryEx
      */
     @Override
     public void switchReadOnlyConnection(Boolean mustBeReadOnly) throws QueryException {
-        if (mustBeReadOnly != currentReadOnlyAsked.get() && currentProtocol.inTransaction()) {
+        if (mustBeReadOnly != currentReadOnlyAsked && currentProtocol.inTransaction()) {
             throw new QueryException("Trying to set to read-only mode during a transaction");
         }
         checkWaitingConnection();
-        if (currentReadOnlyAsked.compareAndSet(!mustBeReadOnly, mustBeReadOnly)) {
```

Comment:
Here is where I believe to be the problem...

While the compareAndSet can ensure that the currentReadOnlyAsked is in a consistent state, it does not guard us through the logic contained within this if condition.

Consider if you will, two threads A & B (I am assuming two threads can hit this in parallel?):

Thread A switches the state (for sake of example false -> true)...CAS works fine
Thread A checks currentReadOnlyAsked and gets true, starts doing work to switch to secondary
Thread A is now forced to yield by OS scheduler, or possibly is blocked on proxy.lock (probably more likely)
Thread B switches the state(true -> false)...CAS again works fine, even though Thread A has not finished changing state
Thread B is able to complete update without blocking, very possible since proxy.lock is not a fair lock (and I am NOT recommending it should be)
Thread A now wakes up and switches connection back to read only

I have solved this by creating a memory barrier for this entire section.  However I am not 100% confident that there is not potential for a deadlock around proxy.lock (I don't see one, but proxy.lock is a nasty beast).  I did email @rusher to discuss the complexity of proxy.lock/MariaDbConnection.lock (same lock), so maybe we want to fix this in a different way with more intense locking changes.

Thoughts?